tests/reconfiguration.py

```diff
@@ -596,6 +597,171 @@ def run_join_old_snapshot(args):
                 pass
 
 
+def get_current_config(network):
+    tables, _ = network.get_latest_ledger_public_state()
+    nwcfgs_tn = "public:ccf.gov.nodes.network.configurations"
+    if nwcfgs_tn not in tables:
+        return None
+    nwcfgs = tables[nwcfgs_tn]
+    if ccf.ledger.WELL_KNOWN_SINGLETON_TABLE_KEY not in nwcfgs:
+        return None
+    current_ptr = json.loads(nwcfgs[ccf.ledger.WELL_KNOWN_SINGLETON_TABLE_KEY])
+    rid = current_ptr["rid"].to_bytes(8, "little")
+    if rid not in nwcfgs:
+        return None
+    current_cfg = json.loads(nwcfgs[rid])
+    return current_cfg
+
+
+def get_current_nodes_table(network):
+    tables, _ = network.get_latest_ledger_public_state()
+    tn = "public:ccf.gov.nodes.info"
+    r = {}
+    for nid, info in tables[tn].items():
+        r[nid.decode()] = json.loads(info)
+    return r
+
+
+def get_joining_node_dates(network, tx_id_before, tx_id_after, node_id):
+    pending_at = 0
+    learner_at = 0
+    trusted_at = 0
+
+    for i in range(tx_id_before.seqno, tx_id_after.seqno + 1):
+        tables = network.get_ledger_public_state_at(i)
+        ntn = "public:ccf.gov.nodes.info"
+        if ntn in tables:
+            for k, v in tables[ntn].items():
+                if k.decode() == node_id:
+                    jv = json.loads(v)
+                    if "status" in jv:
+                        if jv["status"] == "Pending":
+                            pending_at = i
+                        elif jv["status"] == "Learner":
+                            learner_at = i
+                        elif jv["status"] == "Trusted":
+                            trusted_at = i
+
+    return pending_at, learner_at, trusted_at
+
+
+def run_migration_tests(args):
+    if args.reconfiguration_type != "1tx":
+        return
+
+    txs = app.LoggingTxs("user0")
+
+    with infra.network.network(
+        args.nodes,
+        args.binary_dir,
+        args.debug_nodes,
+        args.perf_nodes,
+        pdb=args.pdb,
+        txs=txs,
+    ) as network:
+        network.start_and_join(args)
+        primary, _ = network.find_primary()
+
+        config_before = get_current_config(network)
```

Comment:
In our end-to-end tests, we have found that it is better (as in faster and less flaky) to ask the live service rather than reading the ledger (especially when nodes are still writing to it). For nodes, you can use the existing `/node/network/nodes` endpoint. I don't know if we have an endpoint that returns the current service configuration but we could add it to the existing `/node/network` endpoint (or add a new `/node/service` endpoint)?

Reply:
Thanks, that's good to know! In this particular case I want to make sure that some things are correctly recorded on the ledger, e.g. find the transaction at which a learner becomes trusted, which I figured is best done by reading the ledger directly. I agree though, we should add an endpoint to get the service config.

Reply:
https://microsoft.github.io/CCF/main/operations/operator_rpc_api.html#get--config
https://microsoft.github.io/CCF/main/operations/operator_rpc_api.html#get--consensus

Reply:
That's the node/consensus config, but there is no endpoint for the service config (as in `src/node/config.h`), right?

Reply:
Indeed no, but to find out if a learner is now trusted, that should work. We can and should add an endpoint for the service config, I agree.